packages/wallets/backpack/src/adapter.ts

```diff
@@ -0,0 +1,259 @@
+
+    private _connecting: boolean;
+    private _wallet: BackpackWallet | null;
+    private _publicKey: PublicKey | null;
+    private _readyState: WalletReadyState =
+        typeof window === 'undefined' || typeof document === 'undefined'
+            ? WalletReadyState.Unsupported
+            : WalletReadyState.NotDetected;
+
+    constructor(_config: BackpackWalletAdapterConfig = {}) {
+        super();
+        this._connecting = false;
+        this._wallet = null;
+        this._publicKey = null;
+
+        if (this._readyState !== WalletReadyState.Unsupported) {
+            scopePollingDetectionStrategy(() => {
+                if (window.backpack?.isBackpack) {
+                    this._readyState = WalletReadyState.Installed;
+                    this.emit('readyStateChange', this._readyState);
+                    return true;
+                }
+                return false;
+            });
+        }
+    }
+
+    get publicKey(): PublicKey | null {
+        return this._publicKey;
+    }
+
+    get connecting(): boolean {
+        return this._connecting;
+    }
+
+    get connected(): boolean {
+        return !!this._wallet?.isConnected;
+    }
+
+    get readyState(): WalletReadyState {
+        return this._readyState;
+    }
+
+    async connect(): Promise<void> {
+        try {
+            if (this.connected || this.connecting) return;
+            if (this._readyState !== WalletReadyState.Installed) throw new WalletNotReadyError();
+            this._connecting = true;
+
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            const wallet = window!.backpack!;
+
+            if (!wallet.isConnected) {
+                try {
+                    await new Promise<void>((resolve, reject) => {
+                        const handleConnect = () => {
+                            wallet.off('connect', handleConnect);
+                            resolve();
+                        };
+
+                        // Backpack emits a connect event when it is ready
+                        wallet.on('connect', handleConnect);
+
+                        wallet.connect().catch(() => {
+                            wallet.off('connect', handleConnect);
+                            reject(new WalletWindowClosedError());
+                        });
+                    });
+                } catch (error: any) {
+                    if (error instanceof WalletError) throw error;
+                    throw new WalletConnectionError(error?.message, error);
+                }
+            }
```

Comment:
This code looks like it comes from Phantom's adapter, which now has a warning: https://github.com/solana-labs/wallet-adapter/blob/44a7d3b4910e2ae175ddfef6799a168999fd6ec3/packages/wallets/phantom/src/adapter.ts#L119-L123

If Backpack correctly fulfills the Promise, it can probably be substantially simplified, e.g. https://github.com/solana-labs/wallet-adapter/blob/44a7d3b4910e2ae175ddfef6799a168999fd6ec3/packages/wallets/glow/src/adapter.ts#L130-L134

Reply:
This promise is correctly filled. @tomlinton to confirm.

Reply:
Cool. Yeah, as long as it throws when there's an issue (e.g. if a user declines the connection) then the events shouldn't be needed (here, anyway).

Reply:
It's fulfilled correct in failure cases, but we'll clean up the success case and simplify to the example you linked

Reply:
Ah, okay. Yeah, I guess in the failure case (like if you return instead of throwing if the user declines) we could just check here for the pubkey and throw after. Your call -- I don't want to dictate how Backpack works, just describing how most wallets handle this.